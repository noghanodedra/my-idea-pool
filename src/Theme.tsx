import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

// Define the `theme` prop

export const theme = {
  main: {
    color: {
      primary: "#00A843",
      inputBorder: "#455E70",
      placeholder: "#2a3842",
    },
  },
};

interface IProps {
  children: ReactNode;
}

const Theme = ({ children }: IProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export default Theme;
