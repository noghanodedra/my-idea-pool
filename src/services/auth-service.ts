import { AUTH_DETAILS } from 'constants/common.constants';
import { getAuthDetails } from 'helpers/auth-header';

import API from '../api';

const RESOURCE_NAME = "/access-tokens";

const refreshToken = async (refreshToken: string) => {
    return await API.post(`${RESOURCE_NAME}/refresh`, {
      refresh_token: refreshToken,
    });
}

const login = async (email: string, password: string) => {
  return await API.post(`${RESOURCE_NAME}`, {
    email,
    password,
  });
};

const logout = async (refreshToken: string) => {
  const options = {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      "X-Access-Token": `${getAuthDetails().jwt}`,
    },
    body: JSON.stringify({ "refresh_token": refreshToken }),
  };
  const response = await fetch(
    `${process.env.REACT_APP_API_ENDPOINT}${RESOURCE_NAME}`,
    options
  );
  sessionStorage.removeItem(AUTH_DETAILS);
  return await response.json();
};

export const authService = {
  login,
  logout,
  refreshToken,
};