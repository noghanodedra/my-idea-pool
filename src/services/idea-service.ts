import API from '../api';

const RESOURCE_NAME = "/ideas";

const createIdea = async (
  content: string,
  impact: number,
  ease: number,
  confidence: number
) => {
  return await API.post(`${RESOURCE_NAME}`, {
    content,
    impact,
    ease,
    confidence,
  });
};

const updateIdea = async (
  id: string,
  content: string,
  impact: number,
  ease: number,
  confidence: number
) => {
  return await API.put(`${RESOURCE_NAME}/${id}`, {
    content,
    impact,
    ease,
    confidence,
  });
};

const getIdeas = async (page: number) => {
  return await API.get(`${RESOURCE_NAME}`, { params: { page } });
};

const deleteIdea = async (id: string) => {
  return await API.delete(`${RESOURCE_NAME}/${id}`);
};

export const ideaService = {
  createIdea,
  updateIdea,
  getIdeas,
  deleteIdea,
};
