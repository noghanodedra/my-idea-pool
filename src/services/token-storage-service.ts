export class TokenStorage {
  private static readonly SESSION_STORAGE_TOKEN = "token";
  private static readonly SESSION_STORAGE_REFRESH_TOKEN = "refresh_token";

  public static isAuthenticated(): boolean {
    return this.getToken() !== null;
  }

  public static getAuthentication() {
    return { 
      "X-Access-Token": this.getToken(),
      "Content-Type": "application/json"
    };
  }


  public static storeToken(token: string): void {
    sessionStorage.setItem(TokenStorage.SESSION_STORAGE_TOKEN, token);
  }

  public static storeRefreshToken(refreshToken: string): void {
    sessionStorage.setItem(
      TokenStorage.SESSION_STORAGE_REFRESH_TOKEN,
      refreshToken
    );
  }

  public static clear(): void {
    sessionStorage.removeItem(TokenStorage.SESSION_STORAGE_TOKEN);
    sessionStorage.removeItem(TokenStorage.SESSION_STORAGE_REFRESH_TOKEN);
  }

  public static getRefreshToken(): string | null {
    return sessionStorage.getItem(TokenStorage.SESSION_STORAGE_REFRESH_TOKEN);
  }

  public static getToken(): string | null {
    return sessionStorage.getItem(TokenStorage.SESSION_STORAGE_TOKEN);
  }
}
