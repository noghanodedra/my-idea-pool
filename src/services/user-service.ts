import API from '../api';

const me = async () => {
    return await API.get(`/me`);
}

const signUp = async (email: string, name: string, password: string) => {
  return await API.post(`/users`, {
    email,
    name,
    password,
  });
};

export const userService = {
  me,
  signUp,
};