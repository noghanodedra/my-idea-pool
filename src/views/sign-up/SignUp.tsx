import { EmailValidation, PasswordValidation } from 'components/Common';
import React from 'react';
import { useForm } from 'react-hook-form';
import { useHistory } from 'react-router-dom';
import { TokenStorage } from 'services/token-storage-service';
import { userService } from 'services/user-service';
import styled from 'styled-components';

import {
  DefaultStyledButtonContainer,
  DefaultStyledForm,
  DefaultStyledFormContainer,
  DefaultStyledFormHeader,
  DefaultStyledInput,
  DefaultStyledLink,
  DefaultStyledPrimaryButton,
} from '../../components/BaseComponents';

const StyledFormHeader = styled(DefaultStyledFormHeader)`
  margin-top: 217px;
`;

const StyledButtonContainer = styled(DefaultStyledButtonContainer)`
  margin-top: 20px;
`;

const SignUp = () => {
  const { register, handleSubmit, errors } = useForm();

  const history = useHistory();

  const onSignUp = async (data: any) => {
    try {
      const response = await userService.signUp(
        data.email,
        data.name,
        data.password
      );
      TokenStorage.clear();
      TokenStorage.storeRefreshToken(response.data.refresh_token);
      TokenStorage.storeToken(response.data.jwt);
      history.push("/ideas");
    } catch (e) {
    }
  };

  return (
    <DefaultStyledFormContainer>
      <StyledFormHeader>Sign Up</StyledFormHeader>
      <DefaultStyledForm method="post" onSubmit={handleSubmit(onSignUp)}>
        <DefaultStyledInput
          type="text"
          name="name"
          id="name"
          placeholder="Name"
          aria-label="Full name"
          aria-invalid={errors.name ? "true" : "false"}
          ref={register({
            required: true,
          })}
        />
        {errors.name && errors.name.type === "required" && (
          <p role="alert" className="invalid">
            This is required
          </p>
        )}
        <DefaultStyledInput
          type="email"
          name="email"
          id="email"
          placeholder="Email"
          aria-label="Email"
          aria-invalid={errors.email ? "true" : "false"}
          ref={register(EmailValidation())}
        />
        {errors.email && errors.email.type === "required" && (
          <p role="alert" className="invalid">
            This is required
          </p>
        )}
        {errors.email && errors.email.message && (
          <p role="alert" className="invalid">
            {errors.email.message}
          </p>
        )}
        <DefaultStyledInput
          type="password"
          name="password"
          id="password"
          placeholder="Password"
          aria-label="Password"
          aria-invalid={errors.password ? "true" : "false"}
          ref={register(PasswordValidation())}
        />
        {errors.password && errors.password.type === "required" && (
          <p role="alert" className="invalid">
            This is required
          </p>
        )}
        {errors.password && errors.password.message && (
          <p role="alert" className="invalid">
            {errors.password.message}
          </p>
        )}
        <StyledButtonContainer>
          <DefaultStyledPrimaryButton role="button" type="submit">
            SIGN UP
          </DefaultStyledPrimaryButton>
          <div>
            Already have account ?
            <DefaultStyledLink onClick={() => history.replace("/login")}>
              &nbsp;Log in
            </DefaultStyledLink>
          </div>
        </StyledButtonContainer>
      </DefaultStyledForm>
    </DefaultStyledFormContainer>
  );
};

export default SignUp;
