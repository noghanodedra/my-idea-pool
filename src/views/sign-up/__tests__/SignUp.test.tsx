import 'jest-styled-components';
import 'mutationobserver-shim';

import { fireEvent, render, screen, waitForElement } from '@testing-library/react';
import { shallow } from 'enzyme';
import React from 'react';
import { TokenStorage } from 'services/token-storage-service';
import { userService } from 'services/user-service';
import SignUp from 'views/sign-up/SignUp';

import Theme from '../../../Theme';

const renderComponent = () =>
  render(
    <Theme>
      <SignUp />
    </Theme>
  );

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe("Sign Up Component", () => {
  beforeEach(() => {
    renderComponent();
  });

  it("renders correctly", async () => {
    expect(await screen.findAllByText(/Sign Up/i)).toHaveLength(2);
    expect(await screen.findAllByText(/Log in/i)).toHaveLength(1);
  });

  it("is styled correctly", () => {
    const wrapper = shallow(
      <Theme>
        <SignUp />
      </Theme>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should display required error when value is invalid", async () => {
    fireEvent.submit(screen.getByRole("button"));
    expect(await screen.findAllByRole("alert")).toHaveLength(3);
  });

  it("should display matching error when name is invalid", async () => {
    const emailValue = "test@test.com";
    const passwordValue = "Test@1234";
    const nameValue = "";

    fireEvent.input(screen.getByPlaceholderText("Name"), {
      target: {
        value: nameValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Email"), {
      target: {
        value: emailValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Password"), {
      target: {
        value: passwordValue,
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(screen.getByPlaceholderText("Email")).toHaveValue(emailValue);
    expect(screen.getByRole("textbox", { name: /name/i })).toHaveValue(
      nameValue
    );
    expect(screen.getByPlaceholderText("Password")).toHaveValue(passwordValue);
  });

  it("should display matching error when email is invalid", async () => {
    const emailValue = "test";
    const passwordValue = "Test@1234";
    const nameValue = "test";

    fireEvent.input(screen.getByPlaceholderText("Name"), {
      target: {
        value: nameValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Email"), {
      target: {
        value: emailValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Password"), {
      target: {
        value: passwordValue,
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(screen.getByPlaceholderText("Email")).toHaveValue(emailValue);
    expect(screen.getByRole("textbox", { name: /name/i })).toHaveValue(
      nameValue
    );
    expect(screen.getByPlaceholderText("Password")).toHaveValue(passwordValue);
  });

  it("should display matching error when password is invalid", async () => {
    const emailValue = "test@test.com";
    const passwordValue = "easy";
    const nameValue = "test";

    fireEvent.input(screen.getByPlaceholderText("Name"), {
      target: {
        value: nameValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Email"), {
      target: {
        value: emailValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Password"), {
      target: {
        value: passwordValue,
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(screen.getByPlaceholderText("Email")).toHaveValue(emailValue);
    expect(screen.getByRole("textbox", { name: /name/i })).toHaveValue(
      nameValue
    );
    expect(screen.getByPlaceholderText("Password")).toHaveValue(passwordValue);
  });

  it("should display no error when all fields are valid", async () => {
    const emailValue = "test@test.com";
    const passwordValue = "Test@124";
    const nameValue = "test";

    userService.signUp = jest.fn().mockImplementation((a) => ({
      data: { refresh_token: "test", jwt: "test" },
    }));
    TokenStorage.clear = jest.fn();
    TokenStorage.storeRefreshToken = jest.fn().mockImplementation((a) => a);
    TokenStorage.storeToken = jest.fn().mockImplementation((a) => a);

    fireEvent.input(screen.getByPlaceholderText("Name"), {
      target: {
        value: nameValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Email"), {
      target: {
        value: emailValue,
      },
    });

    fireEvent.input(screen.getByPlaceholderText("Password"), {
      target: {
        value: passwordValue,
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    try {
      await waitForElement(() =>
        expect(screen.queryAllByRole("alert")).toHaveLength(0)
      );
    } catch (error) {}
    expect(screen.getByPlaceholderText("Email")).toHaveValue(emailValue);
    expect(screen.getByRole("textbox", { name: /name/i })).toHaveValue(
      nameValue
    );
    expect(screen.getByPlaceholderText("Password")).toHaveValue(passwordValue);
    expect(TokenStorage.clear).toHaveBeenCalled();
    expect(TokenStorage.storeRefreshToken).toHaveBeenCalled();
    expect(TokenStorage.storeToken).toHaveBeenCalled();
    expect(userService.signUp).toHaveBeenCalled();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
});
