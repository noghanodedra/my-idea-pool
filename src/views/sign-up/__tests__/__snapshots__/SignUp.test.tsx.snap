// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up Component is styled correctly 1`] = `
<ThemeProvider
  theme={
    Object {
      "main": Object {
        "color": Object {
          "inputBorder": "#455E70",
          "placeholder": "#2a3842",
          "primary": "#00A843",
        },
      },
    }
  }
>
  <SignUp />
</ThemeProvider>
`;
